In [2]:
import pandas as pd

# Load the data
df1 = pd.read_csv("../data/raw/df_final_demo.txt")
df2 = pd.read_csv("../data/raw/df_final_experiment_clients.txt")
df3 = pd.read_csv("../data/raw/df_final_web_data_pt_1.txt")
df4 = pd.read_csv("../data/raw/df_final_web_data_pt_2.txt")

# Cleaning columns names
df1.columns = df1.columns.str.lower().str.replace(' ', '_')
df2.columns = df2.columns.str.lower().str.replace(' ', '_')
df3.columns = df3.columns.str.lower().str.replace(' ', '_')
df4.columns = df4.columns.str.lower().str.replace(' ', '_')

# Merge df3 et df4
df_web = pd.concat([df3, df4], axis=0, ignore_index=True)

# Display info
print("===== DF1 (Demographics) =====")
print(f"lines: {len(df1)}, columns: {len(df1.columns)}")
print(f"columns: {df1.columns.tolist()}")
print(f"missing values:\n{df1.isnull().sum()}\n")

print("===== DF2 (Experiment) =====")
print(f"lines: {len(df2)}, columns: {len(df2.columns)}")
print(f"columns: {df2.columns.tolist()}")
print(f"missing values:\n{df2.isnull().sum()}\n")

print("===== DF_WEB(Web Data) =====")
print(f"lines: {len(df_web)}, columns: {len(df_web.columns)}")
print(f"columns: {df_web.columns.tolist()}")
print(f"missing values:\n{df_web.isnull().sum()}")

===== DF1 (Demographics) =====
lines: 70609, columns: 9
columns: ['client_id', 'clnt_tenure_yr', 'clnt_tenure_mnth', 'clnt_age', 'gendr', 'num_accts', 'bal', 'calls_6_mnth', 'logons_6_mnth']
missing values:
client_id            0
clnt_tenure_yr      14
clnt_tenure_mnth    14
clnt_age            15
gendr               14
num_accts           14
bal                 14
calls_6_mnth        14
logons_6_mnth       14
dtype: int64

===== DF2 (Experiment) =====
lines: 70609, columns: 2
columns: ['client_id', 'variation']
missing values:
client_id        0
variation    20109
dtype: int64

===== DF_WEB(Web Data) =====
lines: 755405, columns: 5
columns: ['client_id', 'visitor_id', 'visit_id', 'process_step', 'date_time']
missing values:
client_id       0
visitor_id      0
visit_id        0
process_step    0
date_time       0
dtype: int64


### Checking df1

In [10]:
print(df1.head())
print("=================================================================")
print(df1.info())
print("=================================================================")
print(df1.describe)
print("=================================================================")
print(df1.duplicated().sum())

   client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
0     836976             6.0              73.0      60.5     U        2.0   
1    2304905             7.0              94.0      58.0     U        2.0   
2    1439522             5.0              64.0      32.0     U        2.0   
3    1562045            16.0             198.0      49.0     M        2.0   
4    5126305            12.0             145.0      33.0     F        2.0   

         bal  calls_6_mnth  logons_6_mnth  
0   45105.30           6.0            9.0  
1  110860.30           6.0            9.0  
2   52467.79           6.0            9.0  
3   67454.65           3.0            6.0  
4  103671.75           0.0            3.0  
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70609 entries, 0 to 70608
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   client_id         70609 non-null  int64  
 1   clnt_tenure_yr    7

### Checking df2

In [11]:
print(df2.head())
print("=================================================================")
print(df2.info())
print("=================================================================")
print(df2.describe)
print("=================================================================")
print(df2.duplicated().sum())

   client_id variation
0    9988021      Test
1    8320017      Test
2    4033851   Control
3    1982004      Test
4    9294070   Control
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70609 entries, 0 to 70608
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   client_id  70609 non-null  int64 
 1   variation  50500 non-null  object
dtypes: int64(1), object(1)
memory usage: 1.1+ MB
None
<bound method NDFrame.describe of        client_id variation
0        9988021      Test
1        8320017      Test
2        4033851   Control
3        1982004      Test
4        9294070   Control
...          ...       ...
70604    2443347       NaN
70605    8788427       NaN
70606     266828       NaN
70607    1266421       NaN
70608    9895983       NaN

[70609 rows x 2 columns]>
0


### Checking df_web

In [12]:
print(df_web.head())
print("=================================================================")
print(df_web.info())
print("=================================================================")
print(df_web.describe)
print("=================================================================")
print(df_web.duplicated().sum())  

   client_id            visitor_id                      visit_id process_step  \
0    9988021  580560515_7732621733  781255054_21935453173_531117       step_3   
1    9988021  580560515_7732621733  781255054_21935453173_531117       step_2   
2    9988021  580560515_7732621733  781255054_21935453173_531117       step_3   
3    9988021  580560515_7732621733  781255054_21935453173_531117       step_2   
4    9988021  580560515_7732621733  781255054_21935453173_531117       step_3   

             date_time  
0  2017-04-17 15:27:07  
1  2017-04-17 15:26:51  
2  2017-04-17 15:19:22  
3  2017-04-17 15:19:13  
4  2017-04-17 15:18:04  
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 755405 entries, 0 to 755404
Data columns (total 5 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   client_id     755405 non-null  int64 
 1   visitor_id    755405 non-null  object
 2   visit_id      755405 non-null  object
 3   process_step  755405 non-null  ob

### CREATE A COPY

In [14]:
df1_clean = df1.copy()
df2_clean = df2.copy()
df_web_clean = df_web.copy()

## CLEANING DF1 

In [16]:
missing_rows = df1_clean[df1_clean.isnull().any(axis=1)]
print(missing_rows.head(10))

       client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
4164     7402828             NaN               NaN       NaN   NaN        NaN   
8316      355337             NaN               NaN       NaN   NaN        NaN   
8677     8412164             NaN               NaN       NaN   NaN        NaN   
9583     4666211             8.0             106.0       NaN     F        2.0   
13444    2222915             NaN               NaN       NaN   NaN        NaN   
18066    4876926             NaN               NaN       NaN   NaN        NaN   
25961    5277910             NaN               NaN       NaN   NaN        NaN   
28432    7616759             NaN               NaN       NaN   NaN        NaN   
35323    8191345             NaN               NaN       NaN   NaN        NaN   
43518    1227228             NaN               NaN       NaN   NaN        NaN   

            bal  calls_6_mnth  logons_6_mnth  
4164        NaN           NaN            NaN  
8316        Na

In [17]:
print("GENDR:")
print(df1_clean['gendr'].value_counts(dropna=False))

print("\n\nNUM_ACCTS:")
print(df1_clean['num_accts'].value_counts(dropna=False))

print("\n\nCLNT_AGE (stats):")
print(df1_clean['clnt_age'].describe())

print("\n\nCLNT_TENURE_YR (stats):")
print(df1_clean['clnt_tenure_yr'].describe())

print("\n\nBAL (stats):")
print(df1_clean['bal'].describe())

print("\n\nCALLS_6_MNTH (stats):")
print(df1_clean['calls_6_mnth'].describe())

print("\n\nLOGONS_6_MNTH (stats):")
print(df1_clean['logons_6_mnth'].describe())

GENDR:
gendr
U      24122
M      23724
F      22746
NaN       14
X          3
Name: count, dtype: int64


NUM_ACCTS:
num_accts
2.0    55497
3.0    12529
4.0     2241
5.0      284
6.0       33
NaN       14
7.0        8
1.0        2
8.0        1
Name: count, dtype: int64


CLNT_AGE (stats):
count    70594.000000
mean        46.442240
std         15.591273
min         13.500000
25%         32.500000
50%         47.000000
75%         59.000000
max         96.000000
Name: clnt_age, dtype: float64


CLNT_TENURE_YR (stats):
count    70595.000000
mean        12.052950
std          6.871819
min          2.000000
25%          6.000000
50%         11.000000
75%         16.000000
max         62.000000
Name: clnt_tenure_yr, dtype: float64


BAL (stats):
count    7.059500e+04
mean     1.474452e+05
std      3.015087e+05
min      1.378942e+04
25%      3.734683e+04
50%      6.333290e+04
75%      1.375449e+05
max      1.632004e+07
Name: bal, dtype: float64


CALLS_6_MNTH (stats):
count    70595.000000
m

In [18]:
print(df1_clean[df1_clean.isnull().any(axis=1)])

       client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
4164     7402828             NaN               NaN       NaN   NaN        NaN   
8316      355337             NaN               NaN       NaN   NaN        NaN   
8677     8412164             NaN               NaN       NaN   NaN        NaN   
9583     4666211             8.0             106.0       NaN     F        2.0   
13444    2222915             NaN               NaN       NaN   NaN        NaN   
18066    4876926             NaN               NaN       NaN   NaN        NaN   
25961    5277910             NaN               NaN       NaN   NaN        NaN   
28432    7616759             NaN               NaN       NaN   NaN        NaN   
35323    8191345             NaN               NaN       NaN   NaN        NaN   
43518    1227228             NaN               NaN       NaN   NaN        NaN   
46076    8611797             NaN               NaN       NaN   NaN        NaN   
47189    5144725            

In [19]:
rows_to_drop = df1_clean[df1_clean.isnull().sum(axis=1) > 5].index
df1_clean = df1_clean.drop(rows_to_drop)

In [20]:
df1_clean['gendr'] = df1_clean['gendr'].replace('X', 'U')

In [21]:
median_age = df1_clean['clnt_age'].median()
df1_clean['clnt_age'] = df1_clean['clnt_age'].fillna(median_age)

In [24]:
print(df1_clean.isnull().sum())

client_id           0
clnt_tenure_yr      0
clnt_tenure_mnth    0
clnt_age            0
gendr               0
num_accts           0
bal                 0
calls_6_mnth        0
logons_6_mnth       0
dtype: int64


In [26]:
df1_clean['clnt_age'].value_counts()

clnt_age
58.5    859
55.5    830
29.5    829
28.5    827
57.5    826
       ... 
96.0      2
14.0      2
94.5      1
95.5      1
13.5      1
Name: count, Length: 165, dtype: int64

In [27]:
print(df1[df1['clnt_age'] % 1 == 0.5]['clnt_age'].value_counts().head(10))

clnt_age
58.5    859
55.5    830
29.5    829
28.5    827
57.5    826
27.5    824
52.5    821
30.5    820
31.5    810
54.5    808
Name: count, dtype: int64


In [28]:
df1_clean = df1_clean.rename(columns={
    'gendr': 'gender',
    'clnt_tenure_yr': 'client_tenure_years',
    'clnt_tenure_mnth': 'client_tenure_months',
    'clnt_age': 'client_age',
    'num_accts': 'num_accounts',
    'bal': 'balance',
    'calls_6_mnth': 'calls_6_months',
    'logons_6_mnth': 'logons_6_months'
})
print(df1_clean.columns.tolist())

['client_id', 'client_tenure_years', 'client_tenure_months', 'client_age', 'gender', 'num_accounts', 'balance', 'calls_6_months', 'logons_6_months']


## CLEANING DF2

In [30]:
missing_rows_df2 = df2_clean[df2_clean.isnull().any(axis=1)]
print(missing_rows_df2.head(10))

       client_id variation
50500    5459747       NaN
50501    8031000       NaN
50502    1847030       NaN
50503    9713157       NaN
50504    7775828       NaN
50505    7644910       NaN
50506    5945107       NaN
50507    2212145       NaN
50508    9755145       NaN
50509    5129355       NaN


In [31]:
df2_clean.isnull().all(axis=1).sum()

np.int64(0)

In [33]:
df2_clean['variation'].value_counts(dropna=False)

variation
Test       26968
Control    23532
NaN        20109
Name: count, dtype: int64

In [34]:
df2_clean['client_id'].isnull().sum()

np.int64(0)

In [36]:
nan_clients = df2[df2['variation'].isna()]['client_id']
nan_clients

50500    5459747
50501    8031000
50502    1847030
50503    9713157
50504    7775828
          ...   
70604    2443347
70605    8788427
70606     266828
70607    1266421
70608    9895983
Name: client_id, Length: 20109, dtype: int64

In [38]:
nan_in_web = df_web[df_web['client_id'].isin(nan_clients)]
nan_in_web

client_id             visitor_id                      visit_id  \
245298    7775828  995150070_44942900577  669119128_68248839401_799567   
245299    7775828  995150070_44942900577  669119128_68248839401_799567   
245300    7775828  995150070_44942900577  669119128_68248839401_799567   
245301    7775828  995150070_44942900577  669119128_68248839401_799567   
245302    7775828  995150070_44942900577  669119128_68248839401_799567   
...           ...                    ...                           ...   
755377    9895983  473024645_56027518531  498981662_93503779869_272484   
755378    9895983  473024645_56027518531  498981662_93503779869_272484   
755379    9895983  473024645_56027518531  498981662_93503779869_272484   
755380    9895983  473024645_56027518531  498981662_93503779869_272484   
755381    9895983  473024645_56027518531  498981662_93503779869_272484   

       process_step            date_time  
245298      confirm  2017-03-30 01:31:25  
245299       step_3  2017-03-30 01:30:18  
245300       step_2  2017-03-30 01:29:21  
245301       step_1  2017-03-30 01:28:18  
245302        start  2017-03-30 01:28:10  
...             ...                  ...  
755377       step_3  2017-06-15 19:52:09  
755378       step_2  2017-06-15 19:50:37  
755379       step_1  2017-06-15 19:50:05  
755380        start  2017-06-15 19:50:00  
755381        start  2017-06-15 19:44:05  

[128522 rows x 5 columns]

In [39]:
nan_in_web['client_id'].nunique()

20109

In [40]:
nan_in_demo = nan_clients.isin(df1_clean['client_id']).sum()
nan_in_demo

np.int64(20107)

The 20,109 clients with NaN in 'variation' are real, active clients who used the website but were NOT part of the A/B test experiment

For the purpose of this A/B test analysis (Test vs Control) we will REMOVE these clients and keep only experiment participants.

In [41]:
# Apply cleaning: Keep only Test and Control groups
df2_clean = df2[df2['variation'].notna()].copy()

In [42]:
display(df2_clean)

client_id variation
0        9988021      Test
1        8320017      Test
2        4033851   Control
3        1982004      Test
4        9294070   Control
...          ...       ...
50495     393005   Control
50496    2908510   Control
50497    7230446      Test
50498    5230357      Test
50499    6334360      Test

[50500 rows x 2 columns]

In [43]:
df2_clean['variation'].value_counts()

variation
Test       26968
Control    23532
Name: count, dtype: int64

## CLEANING DF_web

In [45]:
df_web.isnull().sum()

client_id       0
visitor_id      0
visit_id        0
process_step    0
date_time       0
dtype: int64

In [47]:
df_web.duplicated().sum()

np.int64(10764)

In [49]:
duplicated_rows = df_web[df_web.duplicated(keep=False)].sort_values(['client_id', 'date_time'])
duplicated_rows

client_id             visitor_id                      visit_id  \
631355       1531  934069404_34543643308  110350144_67985219545_904677   
631356       1531  934069404_34543643308  110350144_67985219545_904677   
631353       1531  934069404_34543643308  110350144_67985219545_904677   
631354       1531  934069404_34543643308  110350144_67985219545_904677   
639529       2078  585735301_21309149782  900598259_99565669243_552843   
...           ...                    ...                           ...   
256562    9997801   500453195_9636586711  829652786_55390211114_665695   
119532    9998346  292425655_16607136645  189177304_69869411700_783154   
119534    9998346  292425655_16607136645  189177304_69869411700_783154   
406814    9998447  747535871_38029188908  309540451_47388672022_606106   
406815    9998447  747535871_38029188908  309540451_47388672022_606106   

       process_step            date_time  
631355        start  2017-06-02 22:32:08  
631356        start  2017-06-02 22:32:08  
631353        start  2017-06-02 22:32:28  
631354        start  2017-06-02 22:32:28  
639529        start  2017-05-06 08:29:00  
...             ...                  ...  
256562        start  2017-04-28 10:59:18  
119532       step_3  2017-03-29 15:37:28  
119534       step_3  2017-03-29 15:37:28  
406814        start  2017-06-03 19:17:15  
406815        start  2017-06-03 19:17:15  

[19816 rows x 5 columns]

In [50]:
print(duplicated_rows[duplicated_rows['client_id'] == 1531934])

Empty DataFrame
Columns: [client_id, visitor_id, visit_id, process_step, date_time]
Index: []


In [51]:
print(duplicated_rows.head(10).to_string())

        client_id             visitor_id                      visit_id process_step            date_time
631355       1531  934069404_34543643308  110350144_67985219545_904677        start  2017-06-02 22:32:08
631356       1531  934069404_34543643308  110350144_67985219545_904677        start  2017-06-02 22:32:08
631353       1531  934069404_34543643308  110350144_67985219545_904677        start  2017-06-02 22:32:28
631354       1531  934069404_34543643308  110350144_67985219545_904677        start  2017-06-02 22:32:28
639529       2078  585735301_21309149782  900598259_99565669243_552843        start  2017-05-06 08:29:00
639530       2078  585735301_21309149782  900598259_99565669243_552843        start  2017-05-06 08:29:00
639524       2078  585735301_21309149782  900598259_99565669243_552843        start  2017-05-06 08:36:18
639525       2078  585735301_21309149782  900598259_99565669243_552843        start  2017-05-06 08:36:18
639521       2078  585735301_21309149782  900598259_995

In [53]:
first_duplicate_client = duplicated_rows.iloc[0]['client_id']
first_duplicate_client

np.int64(1531)

In [54]:
sample = duplicated_rows.head(20)
print(sample[['client_id', 'visit_id', 'process_step', 'date_time']])

        client_id                      visit_id process_step  \
631355       1531  110350144_67985219545_904677        start   
631356       1531  110350144_67985219545_904677        start   
631353       1531  110350144_67985219545_904677        start   
631354       1531  110350144_67985219545_904677        start   
639529       2078  900598259_99565669243_552843        start   
639530       2078  900598259_99565669243_552843        start   
639524       2078  900598259_99565669243_552843        start   
639525       2078  900598259_99565669243_552843        start   
639521       2078  900598259_99565669243_552843        start   
639522       2078  900598259_99565669243_552843        start   
639518       2078  900598259_99565669243_552843        start   
639519       2078  900598259_99565669243_552843        start   
629801       2476   697934864_7073193369_114100        start   
629802       2476   697934864_7073193369_114100        start   
629799       2476   697934864_7073193369

In [56]:
df_web['date_time'].dtype

dtype('O')

In [57]:
df_web_clean['date_time'] = pd.to_datetime(df_web_clean['date_time'])
df_web_clean['date_time']

0        2017-04-17 15:27:07
1        2017-04-17 15:26:51
2        2017-04-17 15:19:22
3        2017-04-17 15:19:13
4        2017-04-17 15:18:04
                 ...        
755400   2017-05-24 18:46:10
755401   2017-05-24 18:45:29
755402   2017-05-24 18:44:51
755403   2017-05-24 18:44:34
755404   2017-06-03 12:17:09
Name: date_time, Length: 755405, dtype: datetime64[ns]

In [58]:
df_web_clean = df_web_clean.drop_duplicates()

In [59]:
df_web_clean.isnull().sum()

client_id       0
visitor_id      0
visit_id        0
process_step    0
date_time       0
dtype: int64

In [60]:
df_web_clean.duplicated().sum()

np.int64(0)

In [61]:
df_web_clean['process_step'].value_counts()

process_step
start      234999
step_1     162797
step_2     132750
step_3     111589
confirm    102506
Name: count, dtype: int64

Context: According to the project instructions, if a customer has several start times, only the LAST one must be considered. This is because customers may abandon the process and restart it multiple times, but we only want to count their final attempt.

#### Step 1: Analyze how many clients have multiple "start" entries

In [63]:
start_data = df_web_clean[df_web_clean['process_step'] == 'start']
clients_multiple_starts = start_data.groupby('client_id').size()

In [64]:
print(f"Total unique clients who started: {len(clients_multiple_starts)}")
print(f"Clients with exactly 1 start: {(clients_multiple_starts == 1).sum()}")
print(f"Clients with 2+ starts: {(clients_multiple_starts > 1).sum()}")
print(f"Maximum starts per client: {clients_multiple_starts.max()}")

Total unique clients who started: 118980
Clients with exactly 1 start: 67618
Clients with 2+ starts: 51362
Maximum starts per client: 66


#### Step 2: Show distribution of starts per client

In [65]:
print(clients_multiple_starts.value_counts().sort_index().head(10))

1     67618
2     25135
3     12024
4      6091
5      3328
6      1687
7      1022
8       634
9       442
10      239
Name: count, dtype: int64


For each client, identify their LAST visit (based on latest start time) and keep all actions from that visit only

In [66]:
# Step 1: Find the last start for each client

last_starts = df_web_clean[df_web_clean['process_step'] == 'start'].groupby('client_id')['date_time'].max()
last_starts

client_id
169       2017-04-12 20:19:36
336       2017-06-01 07:42:43
546       2017-06-17 10:03:29
555       2017-04-15 12:57:56
647       2017-04-12 15:41:28
                  ...        
9999729   2017-05-08 16:08:25
9999768   2017-06-03 18:11:24
9999832   2017-05-16 16:46:03
9999839   2017-03-29 12:11:32
9999875   2017-06-01 22:40:01
Name: date_time, Length: 118980, dtype: datetime64[ns]

In [67]:
# Step 2: For each client, find which visit_id corresponds to their last start we need to merge the last start times with the original data to get the visit_id

last_start_visits = df_web_clean[df_web_clean['process_step'] == 'start'].merge(
    last_starts.reset_index(), 
    on=['client_id', 'date_time'], 
    how='inner'
)[['client_id', 'visit_id']]

In [68]:
last_start_visits.head()

client_id                      visit_id
0    9988021  781255054_21935453173_531117
1    8320017  960651974_70596002104_312201
2    4033851  949661017_22392791362_127391
3    1982004   132494692_73738638511_70823
4    9294070  116640698_27328181095_188045

In [70]:
# ===== VERIFY DATA STRUCTURE =====
print("===== UNDERSTANDING THE DATA =====\n")

# Example: Look at one client with multiple starts
example_client = clients_multiple_starts[clients_multiple_starts > 1].index[0]
print(f"Example client: {example_client}")
print(f"Number of starts: {clients_multiple_starts[example_client]}")

# Show all their actions
client_actions = df_web_clean[df_web_clean['client_id'] == example_client].sort_values('date_time')
print(f"\nAll actions for this client:")
print(client_actions[['client_id', 'visit_id', 'process_step', 'date_time']])

===== UNDERSTANDING THE DATA =====

Example client: 336
Number of starts: 2

All actions for this client:
        client_id                      visit_id process_step  \
628456        336  649044751_80905125055_554468        start   
628331        336  649044751_80905125055_554468        start   

                 date_time  
628456 2017-06-01 07:26:55  
628331 2017-06-01 07:42:43  


In [71]:
# ===== CHECK MORE EXAMPLES =====
print("===== CHECKING MULTIPLE EXAMPLES =====\n")

# Get 5 clients with multiple starts
multiple_start_clients = clients_multiple_starts[clients_multiple_starts > 1].index[:5]

for client in multiple_start_clients:
    print(f"\n--- Client {client} ({clients_multiple_starts[client]} starts) ---")
    client_data = df_web_clean[df_web_clean['client_id'] == client].sort_values('date_time')
    print(client_data[['client_id', 'visit_id', 'process_step', 'date_time']])
    print(f"Unique visit_ids: {client_data['visit_id'].nunique()}")

===== CHECKING MULTIPLE EXAMPLES =====


--- Client 336 (2 starts) ---
        client_id                      visit_id process_step  \
628456        336  649044751_80905125055_554468        start   
628331        336  649044751_80905125055_554468        start   

                 date_time  
628456 2017-06-01 07:26:55  
628331 2017-06-01 07:42:43  
Unique visit_ids: 1

--- Client 722 (2 starts) ---
        client_id                      visit_id process_step  \
245500        722  984487154_55831795985_521110        start   
245499        722  984487154_55831795985_521110       step_1   
245498        722  984487154_55831795985_521110       step_2   
245497        722  984487154_55831795985_521110       step_1   
245496        722  984487154_55831795985_521110        start   
245495        722  984487154_55831795985_521110       step_1   
245494        722  984487154_55831795985_521110       step_2   
245493        722  984487154_55831795985_521110       step_3   
245492        722  984

In [72]:
# Keep only rows that match the client_id + visit_id combinations in last_start_visits
df_web_final = df_web_clean.merge(
    last_start_visits,
    on=['client_id', 'visit_id'],
    how='inner'
)

In [73]:
df_web_clean['client_id'].nunique()

120157

In [74]:
df_web_final['client_id'].nunique()

118980

In [75]:
print(df_web_final['process_step'].value_counts())

process_step
start      185008
step_1     135615
step_2     113718
step_3      97882
confirm     83903
Name: count, dtype: int64


In [76]:
# ===== STEP 4: KEEP ONLY ACTIONS AFTER LAST START WITHIN EACH VISIT =====
print("===== STEP 4: KEEPING ONLY ACTIONS AFTER LAST START =====\n")

# For each client's visit, find the LAST start time
last_start_in_visit = df_web_final[df_web_final['process_step'] == 'start'].groupby(['client_id', 'visit_id'])['date_time'].max().reset_index()
last_start_in_visit.columns = ['client_id', 'visit_id', 'last_start_time']

# Merge this back to df_web_final
df_web_final = df_web_final.merge(last_start_in_visit, on=['client_id', 'visit_id'], how='left')

# Keep only actions that happened AT OR AFTER the last start
df_web_final = df_web_final[df_web_final['date_time'] >= df_web_final['last_start_time']].copy()

# Drop the helper column
df_web_final = df_web_final.drop('last_start_time', axis=1)

print(f"Rows after filtering: {len(df_web_final)}")
print(f"\nProcess steps now:")
print(df_web_final['process_step'].value_counts())

# Verify: Should have exactly ONE start per client now
starts_per_client = df_web_final[df_web_final['process_step'] == 'start'].groupby('client_id').size()
print(f"\nClients with 1 start: {(starts_per_client == 1).sum()}")
print(f"Clients with 2+ starts: {(starts_per_client > 1).sum()}")

===== STEP 4: KEEPING ONLY ACTIONS AFTER LAST START =====

Rows after filtering: 484124

Process steps now:
process_step
start      118980
step_1     100237
step_2      94761
step_3      87318
confirm     82828
Name: count, dtype: int64

Clients with 1 start: 118980
Clients with 2+ starts: 0


In [77]:
# ===== FINAL VERIFICATION =====
print("===== FINAL DF_WEB_CLEAN VERIFICATION =====\n")

print(f"Total rows: {len(df_web_final)}")
print(f"Total unique clients: {df_web_final['client_id'].nunique()}")
print(f"Total unique visits: {df_web_final['visit_id'].nunique()}")

print(f"\nMissing values:")
print(df_web_final.isnull().sum())

print(f"\nProcess step funnel:")
print(df_web_final['process_step'].value_counts())

# Calculate completion rate
starts = df_web_final[df_web_final['process_step'] == 'start']['client_id'].nunique()
confirms = df_web_final[df_web_final['process_step'] == 'confirm']['client_id'].nunique()
completion_rate = (confirms / starts) * 100

print(f"\n--- Completion Rate ---")
print(f"Clients who started: {starts}")
print(f"Clients who completed: {confirms}")
print(f"Completion rate: {completion_rate:.2f}%")

===== FINAL DF_WEB_CLEAN VERIFICATION =====

Total rows: 484124
Total unique clients: 118980
Total unique visits: 118128

Missing values:
client_id       0
visitor_id      0
visit_id        0
process_step    0
date_time       0
dtype: int64

Process step funnel:
process_step
start      118980
step_1     100237
step_2      94761
step_3      87318
confirm     82828
Name: count, dtype: int64

--- Completion Rate ---
Clients who started: 118980
Clients who completed: 74589
Completion rate: 62.69%


In [78]:
df1_clean.to_csv('../data/Clean/df_demo_clean.csv', index=False)
df2_clean.to_csv('../data/Clean/df_experiment_clean.csv', index=False)
df_web_final.to_csv('../data/Clean/df_web_clean.csv', index=False)